internal/enginenetx/DESIGN.md

```diff
@@ -0,0 +1,815 @@
+
+The general idea of this policy is that it depends on:
+
+1. a `*statsManager` that keeps persistent stats about tactics;
+
+2. a "fallback" policy.
+
+In principle, one would expect `LookupTactics` to first return all
+the tactics we can see from the stats and then try tactics obtained
+from the fallback policy. However, this simplified algorithm would
+lead to suboptimal results in the following case:
+
+1. say there are 10 tactics for "api.ooni.io:443" that are bound
+to a specific bridge address that has been discontinued;
+
+2. if we try all these 10 tactics before trying fallback tactics, we
+would waste lots of time failing before falling back.
+
+Conversely, a better strategy is to "remix" tactics as implemented
+by the [mix.go](mix.go) file:
+
+1. we take the first two tactics from the stats;
+
+2. then we take the first four tactics from the fallback;
+
+3. then we remix the rest, not caring much about whether we're
+reading from the stats of from the fallback.
+
+Because we sort tactics from the stats by our understanding of whether
+they are working as intended, we'll prioritize what we know to be working,
+but then we'll also throw some new tactics into the mix. (We read four
+tactics from the fallback because that allows us to include two bridge tactics
+and two DNS tactics, as explained below when we discuss the
+`bridgePolicy` policy.)
+
+### bridgePolicy
+
+The `bridgePolicy` is implemented by [bridgespolicy.go](bridgespolicy.go) and
+rests on the assumptions made explicit above. That is:
+
+1. that there is at least one _bridge_ for "api.ooni.io";
+
+2. that the Web Connectivity Test Helpers accepts any SNI.
+
+Here we're also using the [mix.go](mix.go) algorithm to remix
+two different sources of tactics:
+
+1. the `bridgesTacticsForDomain` only returns tactics for "api.ooni.io"
+using existing knowledge of bridges and random SNIs;
+
+2. the `maybeRewriteTestHelpersTactics` method filters the results
+coming from the fallback tactic such that, if we are connecting
+to a known test-helper domain name, we're trying to hide its SNI.
+
+The first two returned tactics will be bridges tactics for "api.ooni.io",
+if applicable, followed by two tactics generated using the DNS,
+followed by a random remix of all the remaining tactics. This choice of
+returning two and two tactics first, is the
+reason why in `statsPolicy` we return the first four tactics from
+the fallback after getting two tactics from the stats.
+
+## Overall Algorithm
+
+The composed policy is as described in Diagram 1.
+
+Therefore, the compose policy will return the following tactics:
+
+
+1. if there is a `$OONI_HOME/engine/bridges.conf` with a valid entry,
+use it without trying more tactics; otherwise,
+
+2. use the first two tactics coming from stats, if any;
+
+3. then use the first two tactics coming from bridges, if any;
+
+4. then use the first two tactics coming from the DNS, if successful;
```

Comment:
Can we flip the order of these around in the algorithm so that if DNS is working we use that instead of relying on the bridges?

Basically it would be optimal if we first tried everywhere the simple case (i.e. direct DNS resolution without the hardcoded bridge) and then only if that failed we moved onto trying the hardcoded bridge IP.

Reply:
Yes, it is possible to do that by changing the order with which we construct and compose policies.

I will modify the code to this effect.